File: src/python/gluclient/glu_component.py
"""
Definition of the L{GluComponent} class.

This represents a component that resides on a Glu server.
The object is used for introspection (finding out about
the component meta data, the available services and any
parameters) as well as a starting point for resource
creation.

"""

from gluclient.glu_client_exception import GluClientException
from gluclient.glu_parameter        import GluParameter
from gluclient.glu_service          import GluService
from gluclient.glu_resourcetemplate import GluResourceTemplate

class GluComponent(object):
    """
    Represents information about a component on a Glu server.

    This representation can be used by clients to find out about
    component capabilities and also as a starting point to create
    new resources, by utilizing the L{get_resource_template}() function.

    """
    # The keys to the component's meta data dictionary.
    __NAME_KEY                      = "name"
    __DESC_KEY                      = "desc"
    __DOC_KEY                       = "doc"
    __URI_KEY                       = "uri"
    __PARAMS_KEY                    = "params"
    __RCP_KEY                       = "resource_creation_params"
    __RCP_DESC_KEY                  = "desc"
    __RCP_PUBLIC_KEY                = "public"
    __RCP_SUGGESTED_NAME_KEY        = "suggested_name"
    __SERVICES_KEY                  = "services"

    __server                        = None   # Reference to the server on which we reside (GluServer)
    __name                          = None   # Name of this component
    __description                   = None   # Description of this component
    __doc_uri                       = None   # URI of docs for this component
    __doc                           = None   # Docs for this component
    __uri                           = None   # URI of this component
    __parameters                    = None   # Dictionary of parameter definitions
    __rcp_description_param         = None   # Definition of the fixed 'resource description' parameter
    __rcp_public_param              = None   # Definition of the fixed 'resource public' parameter
    __rcp_suggested_name_param      = None   # Definition of the fixed 'suggested name' parameter
    __services                      = None   # Dictionary of service definitions

    def __init__(self, server, cdesc):
        """
        Create a new component representation in memomory.

        @param server:      The Glu server on which the component resides.
        @type server:       L{GluServer}

        @param cdesc:       Dictionary describing the server component. This
                            is the dictionary returned by the server when a
                            component URI is accessed.
        @type cdesc:        dict

        """
        self.__server = server
        try:
            self.__name        = cdesc[self.__NAME_KEY]
            self.__description = cdesc[self.__DESC_KEY]
            self.__doc_uri     = cdesc[self.__DOC_KEY]
            self.__uri         = cdesc[self.__URI_KEY]
            pdict              = cdesc[self.__PARAMS_KEY]
            sdict              = cdesc[self.__SERVICES_KEY]

            # Parse the parameter dictionary and attempt to translate
            # this to a dictionary of proper GluParameter objects.
            self.__parameters = dict()
            for pname, pdef in pdict.items():
                self.__parameters[pname] = GluParameter(pname, pdef)

            # Set the resource creation time parameters, which
            # are always the same for every component. When we
            # create those parameter definitions we specify the
            # dictionary key as the name of the parameter, don't
            # be surprised...
            self.__rcp_description_param    = GluParameter(self.__RCP_DESC_KEY,           cdesc[self.__RCP_KEY][self.__RCP_DESC_KEY])
            self.__rcp_public_param         = GluParameter(self.__RCP_PUBLIC_KEY,         cdesc[self.__RCP_KEY][self.__RCP_PUBLIC_KEY])
            self.__rcp_suggested_name_param = GluParameter(self.__RCP_SUGGESTED_NAME_KEY, cdesc[self.__RCP_KEY][self.__RCP_SUGGESTED_NAME_KEY])

            # Parse the service dictionary and attempt to translate
            # this to a dictionary of proper GluService objects.
            self.__services = dict()
            for sname, sdef in sdict.items():
                self.__services[sname] = GluService(sname, sdef)

        except KeyError:
            raise GluClientException("Server error: Expected key '%s' missing in definition of component '%s'." % (str(e), self.__name))

    def _create_resource(self, rdict):
        """
        Create a new resource for this component on the server.

        Clients should NOT use this method directly. Instead, they
        should use obtain a resource template from the component and
        then call create_resource() on that template.

        @param rdict:   The completed dictionary to be posted to
                        this components URI.
        @type rdict:    dict

        """
        return self.__server._create_resource(self.__uri, rdict)

    def get_name(self):
        """
        Return the name of the component.

        @return:    Name of component.
        @rtype:     string

        """
        return self.__name

    def get_description(self):
        """
        Return the description of the component.

        @return:    Description of the component.
        @rtype:     string

        """
        return self.__description

    def get_docs(self):
        """
        Return the doc string of the component.

        @return:    Documentation of the component.
        @rtype:     string

        """
        if not self.__doc:
            status, self.__doc = self.__server._json_send(self.__doc_uri, status=200)
            
        return self.__doc

    def get_uri(self):
        """
        Return the URI of the component.

        @return:    URI of the component.
        @rtype:     string

        """
        return self.__uri

    def get_server(self):
        """
        Return the GluServer structure of the server on which this component lives.

        @return:    The server of this component.
        @rtype:     L{GluServer}

        """
        return self.__server

    def get_resource_description(self):
        """
        Return parameter definition for resource description parameter.

        @return:    The resource description parameter definition.
        @rtype:     L{GluParameter}

        """
        return self.__rcp_description_param

    def get_resource_public_flag(self):
        """
        Return parameter definition for resource public flag parameter.

        @return:    The resource public flag parameter definition.
        @rtype:     L{GluParameter}

        """
        return self.__rcp_public_param

    def get_resource_suggested_name(self):
        """
        Return parameter definition for resource suggested name parameter.

        @return:    The resource suggested name parameter definition.
        @rtype:     L{GluParameter}

        """
        return self.__rcp_suggested_name_param

    def get_all_parameters(self):
        """
        Return all parameters defined for this component.

        @return:    Dictionary of all parameters.
        @rtype:     dict of L{GluParameter}

        """
        return self.__parameters

    def get_parameter(self, name):
        """
        Return one parameters of this component.

        @param name:    Name of the parameter.
        @type name:     string

        @return:        Dictionary of all parameters.
        @rtype:         L{GluParameter}

        """
        try:
            return self.__parameters[name]
        except KeyError:
            raise GluClientException("Parameter '%s' not defined." % name)

    def get_all_services(self):
        """
        Return all services defined for this component.

        @return:    Dictionary of all services.
        @rtype:     dict of L{GluService}

        """
        return self.__services

    def get_service(self, name):
        """
        Return one service of this component.

        @param name:    Name of the service.
        @type name:     string

        @return:        Dictionary of service definition.
        @rtype:         L{GluService}

        """
        try:
            return self.__services[name]
        except KeyError:
            raise GluClientException("Service '%s' not defined." % name)

    def get_resource_template(self):
        """
        Return a resource template for this component.

        @return:        Resource template.
        @rtype:         L{GluResourceTemplate}

        """
        return GluResourceTemplate(self)

    #
    # For convenience, we offer read access to several
    # elements via properties.
    #
    name        = property(get_name, None)
    docs        = property(get_docs, None)
    description = property(get_description, None)
    uri         = property(get_uri, None)
    server      = property(get_server, None)

File: src/java/org/mulesource/glu/clientapi/GluComponent.java
/*
 * $Id$
 * --------------------------------------------------------------------------------------
 * Copyright (c) MuleSource, Inc.  All rights reserved.  http://www.mulesource.com
 *
 * The software in this package is published under the terms of the CPAL v1.0
 * license, a copy of which has been included with this distribution in the
 * LICENSE.txt file.
 */

package org.mulesource.glu.clientapi;

import java.util.HashMap;

/**
 * Client-side representation of a component on a Glu server.
 * 
 * This can be used by clients to find out about a component
 * and its capabilities. It is also the starting point for
 * creating a new resource that's based on this component.
 */
public class GluComponent
{
    // The well-known URIs where we can find specific server information
    protected static final String            NAME_KEY               = "name";
    protected static final String            DESC_KEY               = "desc";
    protected static final String            DOC_KEY                = "doc";
    protected static final String            URI_KEY                = "uri";
    protected static final String            PARAMS_KEY             = "params";
    protected static final String            RCP_KEY                = "resource_creation_params";
    protected static final String            RCP_DESC_KEY           = "desc";
    protected static final String            RCP_SUGGESTED_NAME_KEY = "suggested_name";
    protected static final String            SERVICES_KEY           = "services";
    
    protected GluServer                      server;
    protected String                         name;
    protected String                         description;
    protected String                         docUri;
    protected String                         doc;
    protected String                         uri;
    protected HashMap<String, GluParameter>  parameters;
    protected GluParameter                   rcpDescriptionParam;
    protected GluParameter                   rcpSuggestedNameParam;
    protected HashMap<String, GluService>    services;
    
    
    /**
     * Create a new GluComponent object, representing a Glu component in the client.
     * 
     * @param   server  Reference to the client-side representation of the Glu server.
     * @param   cdesc   A map with all the information about the component.
     * @throws          GluClientException
     */
    public GluComponent(GluServer server, HashMap<String, ?> cdesc) throws GluClientException
    {
        this.server = server;
        
        // Store the meta data for later use
        try {
            name                     = (String)cdesc.get(NAME_KEY);
            description              = (String)cdesc.get(DESC_KEY);
            docUri                   = (String)cdesc.get(DOC_KEY);
            uri                      = (String)cdesc.get(URI_KEY);
            HashMap<String, ?> pdict = (HashMap<String, ?>)cdesc.get(PARAMS_KEY);
            HashMap<String, ?> sdict = (HashMap<String, ?>)cdesc.get(SERVICES_KEY);

            // Parse the parameter dictionary and attempt to translate
            // this into a dictionary of proper GluParameter objects.
            parameters = new HashMap<String, GluParameter>();
            for (String pname : pdict.keySet()) {
                parameters.put(pname, new GluParameter(pname, (HashMap<String, ?>)pdict.get(pname)));
            }
            
            // Set the resource creation time parameters, which
            // are always the same for every component. When we
            // create those parameter definitions we specify the
            // dictionary key as the name of the parameter, don't
            // be surprised...
            HashMap<String, HashMap<String, ?>> rcpMap = (HashMap<String, HashMap<String, ?>>) cdesc.get(RCP_KEY);
            rcpDescriptionParam   = new GluParameter(RCP_DESC_KEY,           rcpMap.get(RCP_DESC_KEY));
            rcpSuggestedNameParam = new GluParameter(RCP_SUGGESTED_NAME_KEY, rcpMap.get(RCP_SUGGESTED_NAME_KEY));

            // Parse the services dictionary and attempt to translate
            // this into a dictionary of proper GluService objects.
            services = new HashMap<String, GluService>();
            for (String sname : sdict.keySet()) {
                services.put(sname, new GluService(sname, (HashMap<String, ?>)sdict.get(sname)));
            }
        }
        catch (Exception e) {
            System.out.println("Exception: " + e);
            throw new GluClientException("Server error: Could not process component meta data: " + e.getMessage());
        }
    }
    
    /**
     * Sends the request to create a new resource on the server.
     * 
     * Clients don't use this method, but instead create a resource
     * through a {@link GluResourceTemplate} object.
     * 
     * This method sends a request to this components own URI.
     * 
     * @param  rdict The dictionary with all required parameters for the
     *               resource creation.
     *               
     * @return       An HttpResult object, containing the response from the server.
     * 
     * @throws       GluClientException
     */
    protected HashMap<String, Object> createResource(Object rdict) throws GluClientException
    {
        return server.createResource(uri, rdict);
    }
    
    /**
     * Returns the name of this component.
     * 
     * @return  Name of the component.
     */
    public String getName()
    {
        return name;
    }
    
    /**
     * Returns the description for this component.
     * 
     * @return  Description of the component.
     */
    public String getDescription()
    {
        return description;
    }
    
    /**
     * Return the doc page for this component.
     * 
     * This may generate a request to the server for the component's docs.
     * Once we receive the information, it is cached so that we can reused
     * it later on.
     * 
     * @return  Doc page for this component.
     * @throws  GluClientException
     */
    public String getDocs() throws GluClientException
    {
        if (doc == null) {
            HttpResult res = server.jsonSend(docUri, null, null, 200, null);
            doc = (String)res.data;
        }
        return doc;
    }
    
    /**
     * Returns reference to the GluServer for this component.
     * 
     * @return  {@link GluServer} reference.
     */
    public GluServer getServer()
    {
        return server;
    }

    /**
     * Return definition of the description parameter.
     * 
     * The description parameter is used when creating a resource.
     * 
     * @return  Definition of the description parameter.
     */
    public GluParameter getResourceDescription()
    {
        return rcpDescriptionParam;
    }

    /**
     * Return definition of the suggested name parameter.
     * 
     * The suggested name parameter is used when creating a resource.
     * 
     * @return  Definition of the suggested name parameter.
     */
    public GluParameter getSuggestedName()
    {
        return rcpSuggestedNameParam;
    }
    
    /**
     * Return a map of all the parameter definitions for this component.
     * 
     * @return  Map of {@link GluParameter} objects.
     */
    public HashMap<String, GluParameter> getAllParameters()
    {
        return parameters;
    }
    
    /**
     * Return a single parameter definition.
     * 
     * @return  A single {@link GluParameter} object.
     */
    public GluParameter getParameter(String name)
    {
        return parameters.get(name);
    }
    
    /**
     * Return a map of all the service definitions for this component.
     * 
     * @return  Map of {@link GluService} objects.
     */
    public HashMap<String, GluService> getAllServices()
    {
        return services;
    }
    
    /**
     * Returns a service object for the specified service.
     * 
     * @param  name  Name of the service.
     * @return       Object representing this service.
     */
    public GluService getService(String name)
    {
        return services.get(name);
    }
    
    /**
     * Return template for creation of new resource based on this component.
     * 
     * @return  A new resource creation template.
     */
    public GluResourceTemplate getResourceTemplate()
    {
        return new GluResourceTemplate(this);
    }
}


